File: php/src/Exports/Robots.php
<?php
declare(strict_types=1);

namespace Bkbs\Exports;

final class Robots
{
    public const BEGIN = '# BEGIN BKBS';
    public const END = '# END BKBS';

    /**
     * @param array<string, mixed> $site
     */
    public static function contentUsageLine(array $site): ?string
    {
        $search = !empty($site['aipref_search']);
        $aiInput = !empty($site['aipref_ai_input']);
        $train = !empty($site['aipref_train_ai']);
        if (!$search && !$aiInput && !$train) {
            return null;
        }
        $yn = static fn(bool $v): string => $v ? 'y' : 'n';
        return 'Content-Usage: search=' . $yn($search) . ', ai-input=' . $yn($aiInput) . ', train-ai=' . $yn($train);
    }

    /**
     * @param array<string, mixed> $site
     */
    public static function block(array $site): string
    {
        $base = rtrim((string) ($site['base_url'] ?? ''), '/');
        $lines = [
            self::BEGIN,
            '# Machine layers for AI agents (managed by Manifest BKBS Converter)',
            'User-agent: *',
            'Allow: /llms.txt',
            'Allow: /llms-full.txt',
            'Allow: /graph.json',
            'Allow: /schema/',
            'Allow: /.well-known/agent.json',
        ];
        $usage = self::contentUsageLine($site);
        if ($usage !== null) {
            $lines[] = $usage;
        }
        $lines[] = 'Sitemap: ' . $base . '/sitemap.xml';
        $lines[] = self::END;
        return implode("\n", $lines) . "\n";
    }

    /**
     * @param array<string, mixed>|string $siteOrBaseUrl
     */
    public static function merge(string $root, array|string $siteOrBaseUrl): void
    {
        $site = is_array($siteOrBaseUrl)
            ? $siteOrBaseUrl
            : ['base_url' => $siteOrBaseUrl];
        $path = $root . '/robots.txt';
        $block = self::block($site);
        $existing = is_file($path) ? (string) file_get_contents($path) : '';
        $begin = self::BEGIN;
        $end = self::END;
        if (str_contains($existing, $begin)) {
            $startPos = strpos($existing, $begin);
            $endPos = strpos($existing, $end);
            if ($startPos === false) {
                $existing = rtrim($existing) . "\n\n" . $block;
            } else {
                if ($endPos !== false) {
                    $endPos += strlen($end);
                    $after = ltrim(substr($existing, $endPos));
                } else {
                    $after = '';
                }
                $existing = rtrim(substr($existing, 0, $startPos)) . "\n\n" . $block;
                if ($after !== '') {
                    $existing .= $after;
                }
            }
        } else {
            $existing = rtrim($existing) . ($existing !== '' ? "\n\n" : '') . $block;
        }
        $dir = dirname($path);
        if (!is_dir($dir)) {
            mkdir($dir, 0755, true);
        }
        file_put_contents($path, $existing);
    }
}

File: app/exports/robots.py
"""robots.txt suggestion + live merge block (Stage 4c AIPREF opt-in)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

BEGIN = "# BEGIN BKBS"
END = "# END BKBS"


def _yn(flag: bool) -> str:
    return "y" if flag else "n"


def aipref_enabled(site: Any) -> bool:
    return bool(
        getattr(site, "aipref_search", False)
        or getattr(site, "aipref_ai_input", False)
        or getattr(site, "aipref_train_ai", False)
    )


def content_usage_line(site: Any) -> str | None:
    """IETF AIPREF Content-Usage. None when all toggles are off (default)."""
    if not aipref_enabled(site):
        return None
    return (
        "Content-Usage: search="
        + _yn(bool(getattr(site, "aipref_search", False)))
        + ", ai-input="
        + _yn(bool(getattr(site, "aipref_ai_input", False)))
        + ", train-ai="
        + _yn(bool(getattr(site, "aipref_train_ai", False)))
    )


def robots_suggestion(site: Any) -> str:
    return f"""# Suggested robots.txt additions for BKBS / AI agents
# Merge carefully with your existing robots.txt

User-agent: *
Allow: /

# Explicitly allow common AI crawlers if desired
User-agent: GPTBot
Allow: /

User-agent: ClaudeBot
Allow: /

User-agent: Google-Extended
Allow: /

Sitemap: {site.base_url}/sitemap.xml
"""


def managed_robots_block(site: Any) -> str:
    base = str(site.base_url).rstrip("/")
    lines = [
        BEGIN,
        "# Machine layers for AI agents (managed by Manifest BKBS Converter)",
        "User-agent: *",
        "Allow: /llms.txt",
        "Allow: /llms-full.txt",
        "Allow: /graph.json",
        "Allow: /schema/",
        "Allow: /.well-known/agent.json",
    ]
    usage = content_usage_line(site)
    if usage:
        lines.append(usage)
    lines.append(f"Sitemap: {base}/sitemap.xml")
    lines.append(END)
    return "\n".join(lines) + "\n"


def merge_robots(root: Path, site: Any) -> str | None:
    """
    Append or replace the BKBS allow block in robots.txt.
    Always writes # END BKBS. If # BEGIN BKBS exists without END, replace from
    BEGIN to EOF (do not duplicate). Returns relative path written.
    """
    robots = root / "robots.txt"
    block = managed_robots_block(site)
    existing = robots.read_text(encoding="utf-8") if robots.exists() else ""
    robots.parent.mkdir(parents=True, exist_ok=True)
    if BEGIN in existing:
        start = existing.find(BEGIN)
        end = existing.find(END)
        if end >= 0:
            end += len(END)
            after = existing[end:].lstrip("\n")
        else:
            after = ""
        new = existing[:start].rstrip() + "\n\n" + block
        if after:
            new += "\n" + after
        robots.write_text(new.rstrip() + "\n", encoding="utf-8")
    else:
        robots.write_text((existing.rstrip() + "\n\n" + block).lstrip() + "\n", encoding="utf-8")
    return "robots.txt"
